utils/src/main/java/software/amazon/awssdk/utils/async/IterablePublisher.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.utils.async;
+
+import java.util.Iterator;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.utils.Validate;
+
+@SdkProtectedApi
+public class IterablePublisher<T> implements Publisher<T> {
+    private final Object iterableLock = new Object();
+    private final Iterable<T> iterable;
+
+    public IterablePublisher(Iterable<T> iterable) {
+        this.iterable = Validate.paramNotNull(iterable, "iterable");
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber) {
+        Iterator<T> iterator = iterable.iterator();
+        SimplePublisher<T> publisher = new SimplePublisher<>();
+
+        // Prime the simple publisher with 1 event. More will be sent as these complete.
+        sendEvent(iterator, publisher);
+
+        publisher.subscribe(subscriber);
+    }
+
+    private void sendEvent(Iterator<T> iterator, SimplePublisher<T> publisher) {
+        try {
+            synchronized (iterableLock) {
```

Comment:
Because only one element is in flight at a time, I think we can remove this lock now.